### PR TITLE
kubelet: warn instead of error for unsupported options on Windows

### PR DESCRIPTION
**Pull request by neolit123**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Make validateKubeletOSConfiguration() show warnings instead of returning errors on Windows for the fields "CgroupsPerQOS" and "EnforceNodeAllocatable".

the argument here is simple, we don't need to exit the kubelet with an error on Windows just because the user is using a config that works on Linux. We could even take this one step further and silently modify these values on Windows and not alert the user.

#### Which issue(s) this PR fixes:

xref https://github.com/kubernetes/kubeadm/issues/2967
xref https://github.com/kubernetes/kubernetes/pull/123136

old PR where we discussed that different defaults on Windows are not desired:
- https://github.com/kubernetes/kubernetes/pull/77710
- https://github.com/kubernetes/kubernetes/pull/77710#issuecomment-491450781
- 
#### Does this PR introduce a user-facing change?

```release-note
kubelet: warn instead of error for the unsupported options on Windows "CgroupsPerQOS" and "EnforceNodeAllocatable".
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/kubelet/apis/config/validation/validation_windows.go
Comment (bart0sh) [inline pkg/kubelet/apis/config/validation/validation_windows.go:31]: Would it make sense to explicitly call `klow.Warningf` here instead of passing it down the stack from the `kubelet/app/server.go`?

Comment (neolit123) [reply pkg/kubelet/apis/config/validation/validation_windows.go:31]: i did it informally like that, but i can change it if the maintainers prefer to call directly.

i think an API package should not import  klog.

Comment (bart0sh) [reply pkg/kubelet/apis/config/validation/validation_windows.go:31]: It makes sense to me. Thanks for the explanations. Let's see what other reviewers say.

not sure if it makes sense to do in this particular case, but to avoid importing klog you can pass context and get the logger from it using `FromContext`.

**Comment by dims** [comment]
/approve
/lgtm

No logic changes for kubelet itself, just reorganizing code a bit to make things better for windows

**Review by mikebrow — COMMENTED** [review]
Prefer the mentioned silent option (ignoring linux config that has no map but is seen by windows team as ok by them..) over making it a valid config but outputting new noisy warnings. Looks like the soln is to rm the validation_windows.go file..

**Comment by neolit123** [comment]
> Prefer the mentioned silent option (ignoring linux config that has no map but is seen by windows team as ok by them..) over making it a valid config but outputting new noisy warnings. Looks like the soln is to rm the validation_windows.go file..

not a big fan of the warningf pass-around either.
if we are OK to not print warnings i can make validation_windows.go return nil (no errors). it is still required, to abstract between the OSes.

@jsturtevant WDYT?
i think this will make the kubelet maintainers happier.
cc @mrunalp

**Comment by jsturtevant** [comment]
> not a big fan of the warningf pass-around either.

+ 1

> if we are OK to not print warnings i can make validation_windows.go return nil (no errors). i

I would prefer to change the defaults to be OS specific than not print out warnings. My concern is that users will set those values and expect things to work but have no way to of knowing they don't.  With warnings or errors they can figure this out.

**Comment by knabben** [comment]
+1 for showing a warning for not supported flags for Windows.

**Comment by jsturtevant** [comment]
> i can call klog.Warning directly instead of passing the warningf, which doesn't matter to me.

This works too, I guess I am not sure of the difference between the two

/cc @marosset @aravindhp

**Comment by neolit123** [comment]
> > i can call klog.Warning directly instead of passing the warningf, which doesn't matter to me.
> 
> This works too, I guess I am not sure of the difference between the two
> 
> /cc @marosset @aravindhp

it is a smaller diff, but i didn't want to import a logger in the kubelet API package.
if the maintainers are fine with that i can change it.

https://github.com/kubernetes/kubernetes/pull/123137#discussion_r1487541581

**Comment by neolit123** [comment]
updated the PR:
- don't plumb a warningf() around, which is messy
- import klog in options_other.go and call Warningf, this is in the internal API package so importing a logger should be OK, definitely a no-go for external APIs.

File: pkg/kubelet/apis/config/validation/validation_windows.go
Comment (mikebrow) [inline pkg/kubelet/apis/config/validation/validation_windows.go:30]:
```suggestion
	message := "ignored configuration option: %v (%v) %v is not supported on Windows"
```

**Review by mikebrow — COMMENTED** [review]
nit regarding text of warning message.. a-priori this would throw an error and not run

With this change, the options are ignored with the exception of output warning(s).

File: pkg/kubelet/apis/config/validation/validation_windows.go
Comment (neolit123) [reply pkg/kubelet/apis/config/validation/validation_windows.go]: thanks, updated.

**Review by mikebrow — APPROVED** [review]
/lgtm

**Comment by neolit123** [comment]
> Once this PR has been reviewed and has the lgtm label, please assign liggitt

as per the bot recommendation.
/assign @liggitt

**Review by jsturtevant — COMMENTED** [review]
/lgtm

**Comment by pacoxu** [comment]
> This pull-request has been approved by: _[dims](https://github.com/kubernetes/kubernetes/pull/123137#issuecomment-2124489117)_, _[mikebrow](https://github.com/kubernetes/kubernetes/pull/123137#pullrequestreview-2085583291)_, _[neolit123](https://github.com/kubernetes/kubernetes/pull/123137#)_

/assign @mrunalp @derekwaynecarr @yujuhong 
for approval from sig-node

**Comment by neolit123** [comment]
update:
i joined yesterday's sig node meeting a spoke about this PR. @mrunalp promised to have a look, but he also raised the Q of why we are not having different defaults on Windows. my response was that this was discussed on another occasion in the past and it was not approved as an idea:
- https://github.com/kubernetes/kubernetes/pull/77710

**Comment by yujuhong** [comment]
/approve

since we don't have different defaults for windows

**Comment by pacoxu** [comment]
/assign @liggitt 
for api approval

**Comment by neolit123** [comment]
@mrunalp from SIG node during the zoom call today, said that this is waiting on API approval.
@liggitt could you please take a look? thanks

**Pull request merged**
